### PR TITLE
Adds AWS SQS FIFO queues support

### PR DESCRIPTION
**Pull request by neeckeloo**
[AWS SQS supports FIFO Queues](http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/FIFO-queues.html) was introduced by Amazon in order to offer strictly ordered message delivery and exactly-once message processing.
A FIFO queue has a name suffixed by `.fifo` and each message must contains additional attributes:
- `MessageGroupId`: The tag that specifies that a message belongs to a specific message group. Messages that belong to the same message group are processed in a FIFO manner (however, messages in different message groups might be processed out of order).
- `MessageDeduplicationId`: If a message with a particular `MessageDeduplicationId` is sent successfully, any messages sent with the same `MessageDeduplicationId` are accepted successfully but aren't delivered during the 5-minute deduplication interval.

More details in this [blog post](https://aws.amazon.com/blogs/developer/how-the-amazon-sqs-fifo-api-works/).

### CODE REVIEW

**Comment by neeckeloo** [comment]
@basz @OctavianParalescu I had not seen that an issue #45 had just been created. So, what do you think about?

File: src/SlmQueueSqs/Queue/SqsQueue.php [hunk 1]
```diff
@@ -64,6 +66,15 @@ public function push(JobInterface $job, array $options = array())
             'DelaySeconds' => isset($options['delay_seconds']) ? $options['delay_seconds'] : null
         );
 
+        if ($this->isFifoQueue()) {
+            if (!isset($options['message_group_id']) || empty($options['message_group_id'])) {
+                throw new Exception\MissingMessageGroupException();
+            }
+
+            $parameters['MessageGroupId'] = $options['message_group_id'];
+            $parameters['MessageDeduplicationId'] = md5($parameters['MessageBody']);
```
Comment (OctavianParalescu) [inline src/SlmQueueSqs/Queue/SqsQueue.php]: You can already do by default a content-based deduplication only with a setting in AWS thus not requiring that MessageDeduplicationId. 

My opinion is that we should let the user decide if they set the deduplication id because:
- MessageDeduplicationId might not be required if content-based deduplication is enabled for the queue
- Users might want to enforce their deduplication-id-generation technique

File: src/SlmQueueSqs/Queue/SqsQueue.php [hunk 2]
```diff
@@ -262,4 +273,27 @@ public function batchDelete(array $jobs)
...
+    private function endsWith($haystack, $needle)
+    {
+        $length = strlen($needle);
+        if ($length === 0) {
+            return true;
+        }
+
+        return (substr($haystack, -$length) === $needle);
+    }
 }
```
Comment (OctavianParalescu) [inline src/SlmQueueSqs/Queue/SqsQueue.php]: Batch push also needs those parameters.

File: src/SlmQueueSqs/Queue/SqsQueue.php (hunk 1 above)
Comment (neeckeloo) [reply src/SlmQueueSqs/Queue/SqsQueue.php]: I agree with leaving the choice to the user as I did for the `MessageGroupId`.

Comment (OctavianParalescu) [reply src/SlmQueueSqs/Queue/SqsQueue.php]: Cheers! What about batch push?

Comment (neeckeloo) [reply src/SlmQueueSqs/Queue/SqsQueue.php]: I just pushed a commit that also add the parameters to the `batchPush` method.

**Comment by neeckeloo** [comment]
@OctavianParalescu Does it seem relevant to offer the possibility to let the module generate the `MessageDeduplicationId` parameter using `enable_auto_deduplication` key even if its possible by enabling `ContentBasedDeduplication` directly for the queue?

**Comment by OctavianParalescu** [comment]
@neeckeloo If the user would enable auto deduplication, it would mean you will `md5()` or `insertHashFunctionHere()` on the message body. AWS already does exactly that if you enable the automatic content auto deduplication. It applies a `sha512 `(don't remember exactly) to the body of the message.

Although we may add some default functionality to add the deduplication id automatically as by default, fifo queues have that deduplication rule turned off. 

In conclusion we may create a enable_auto_deduplication setting in the config, set it to be true by default and when it is true, set `MessageDeduplicationId `as `md5(body)` if it is not already defined. Great idea!

**Comment by basz** [comment]
Merge conflict

**Comment by neeckeloo** [comment]
Merge conflicts have been resolved.

**Comment by webit4me** [comment]
Really like to have this feature in, is it possible to have it merged?
